lib/Dialect/RTL/Ops.cpp

```diff
@@ -657,12 +657,24 @@ void AddOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
         return success();
       }
 
+      // add(..., x, x) -> add(..., shl(x, 1))
+      if (inputs[size - 1] == inputs[size - 2]) {
```

Comment:
This is a good canonicalization, but please make it be specific to binary adds, or handle the case where x shows up multiple times.  We want `add(x,x,x)` to turn into `mul(x, 3)` for example.

Reply:
This could also be done by teaching the system to transform `add(... x, shl(x,c))` -> add with multiply.  The same sort of thing should also handle multiply operands.

Reply:
I naively just did the TODO, but putting more thought into this, I'm thinking of these three approaches:

1. Just use multiply through and through, and avoid left shifts entirely. 
2. Have the base case of `add(…, x, x) -> add(…, shl(x, 1))` , and then from there if we get another x turn it into multiplies. i.e. `add(…, x, shl(x, 1)) -> add(…, mul(x, 3))`, and continue using multiply.
3. Go back and forth between shift and multiply. If c is a power of two, use shift, where shift amount `s = log2(c)`. Otherwise, use multiply. Here, c is the number of times x appears.

Looking for some input on which approach is the best.

Reply:
I had a second, and implemented the 3rd option. There may be a smarter way to pull this off that I'm just missing here.

Reply:
Ok, yeah there are tradeoffs here :).  the important thing is that the series of canonicalizations compose correctly and "hillclimb" towards a global optima (in terms of canonical form, not necessary "performance").  Approach #3 seems logical to me!

This raises the question of how best to represents shifts by a constant.  Unlike for CPUs, this is a markedly different operation for HW than it is for SW.  It might makes sense to have a specific opcode for it. Modeling it as a bitextract+extend is another possiblity but will make algebraic simplication like this really fragile.  Would you be interested in raising this at the next design meeting?

Reply:
Yeah that's why I wasn't sure what the global optima was in terms of canonical form.

Typically I'd jump at chance like this to embarrass myself. However, CIRCT design meetings are during a lecture for me (except for last Wednesday given that we had the day off).

This is interesting to me however, and something I hadn't thought about, nor really grasp enough to say which is the better choice given my lack of time at the hardware level. I'm really curious how this plays out however.

Reply:
Ok, fair enough, we can keep it to this forum.